libs/reconciler/README.md

```diff
@@ -0,0 +1,475 @@
+
+  1. Prepare `Item` and `Configurator` implementations.
+  2. Build a registry mapping Configurators to items (consider using `DefaultRegistry`).
+  3. Prepare two depgraph instances - one to represent the current state and the other
+     modeling the intended state. If the current state graph is `nil`, it means that
+     there are no items currently created. If the intended state graph is `nil`, it means
+     that all items inside the current state should be deleted.`Reconcile()` will return
+     updated/created current state graph, but it will not change the intended state.\
+     Note that by using subgraphs (see [Readme for depgraph](../depgraph/README.md)),
+     it is possible to select only a "chunk" of the graph and run reconciliation only
+     for that. If you know that only a subset of the state needs to be updated, use subgraph
+     to decrease the size of the input and thus make the reconciliation less expensive.
+     It is even possible to select only a single node for reconciliation.
+
+A great thing about the Reconciler is that it is completely stateless, you can throw
+it out after `Reconcile()` returns. For every reconciliation you can create a new
+instance using `reconciler.New()`, it costs practically nothing. And even if there are
+some async operations running in the background, you can resume the reconciliation with
+a new instance of Reconciler. Any in-progress state information is kept inside the graph
+representing the current state.
+
+Using Reconciler for items with only synchronous state transitions is trivial:
+
+```go
+// Registry should only be created once (during some init procedure).
+registry := &reconciler.DefaultRegistry{}
+// ...
+
+// Running reconciliation when needed with a new instance of Reconciler:
+//  - ctx is of type context.Context
+//  - currentState is of type depgraph.Graph
+//  - intendedState is of type depgraph.GraphR (subset of depgraph.Graph methods)
+// See libs/depgraph/README.md to learn how to build and update a graph.
+r := reconciler.New(registry)
+status := r.Reconcile(ctx, currentState, intendedState)
+if status.Err != nil {
+     fmt.Printf("State reconciliation failed: %v\n", status.Err)
+}
+fmt.Printf("Executed state transitions:\n%s\n", status.OperationLog)
+
+// currentState was updated by Reconciler in-place (no need to take from status.NewCurrentState).
+// However, if currentState was nil, Reconciler created a new graph instance.
+currentState = status.NewCurrentState
+```
+
+With asynchronous operations the usage is still relatively simple.\
+Let's assume that a management agent is built around a single main event loop,
+consisting of an infinite `for` loop and a `select` statement waiting for events.
+Some of these events may result in a change for the intended state and therefore
+require to run a state reconciliation.
+
+Reconciler always only waits for synchronous operations and exists immediately when all
```

Comment:
Did you mean "exits"?

Reply:
So there are two kinds of configurator actions, those that are sync and those that are async?

I find this model quite confusing, probably because it is different than how most promises-based engines I have seen work.

Is there any reason that we cannot do a promise-style? The following is the API we always use:

```go
ch := r.Reconcile(ctx, currentState, intendedState)
```

It always returns a channel, which we can monitor for when things are done. That would simplify the interface from the outside. Whether the configurator actions are sync or async, doesn't matter to what calls `Reconcile()`. It always gets a `channel` and selects on that channel. The returned object can contain an error or errors, success, whatever. The interface is just a promise.

Reply:
> Did you mean "exits"?

Yes, fixed.

> Is there any reason that we cannot do a promise-style?

See my comments above. Mainly I didn't want the Reconciler to run in a separate Go routine. The reason is to avoid race-conditions between successive reconciliations as well as between the Reconciler and the user's code. Basically the user decides what runs in a separate Go routine (inside a Configurator) and what runs synchronously from within `Reconcile()`.